source/detail/numeric_utils.hpp

```diff
@@ -0,0 +1,116 @@
+// Copyright (c) 2014-2018 Thomas Fussell
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, WRISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE
+//
+// @license: http://www.opensource.org/licenses/mit-license.php
+// @author: see AUTHORS file
+
+#pragma once
+
+#include <xlnt/xlnt_config.hpp>
+#include <cmath>
+#include <limits>
+#include <sstream>
+#include <type_traits>
+
+namespace xlnt {
+namespace detail {
+/// <summary>
+/// Takes in any number and outputs a string form of that number which will
+/// serialise and deserialise without loss of precision
+/// </summary>
+template <typename Number>
+std::string serialize_number_to_string(Number num)
+{
+    // more digits and excel won't match
+    constexpr int Excel_Digit_Precision = 15; //sf
+    std::stringstream ss;
+    ss.precision(Excel_Digit_Precision);
+    ss << num;
+    return ss.str();
+}
+
+/// <summary>
+/// constexpr abs
+/// </summary>
+template <typename Number>
+constexpr Number abs(Number val)
+{
+    return (val < Number{0}) ? -val : val;
+};
+
+/// <summary>
+/// constexpr max
+/// </summary>
+template <typename NumberL, typename NumberR>
+constexpr typename std::common_type<NumberL, NumberR>::type max(NumberL lval, NumberR rval)
+{
+    return (lval < rval) ? rval : lval;
+}
+
+/// <summary>
+/// constexpr min
+/// </summary>
+template <typename NumberL, typename NumberR>
+constexpr typename std::common_type<NumberL, NumberR>::type min(NumberL lval, NumberR rval)
+{
+    return (lval < rval) ? lval : rval;
+}
```

Comment:
float_equals could be fully constexpr with c++14. I originally wrote it with constexpr and then removed it later (single statement restriction for cpp11), hence these simple utilities. Could use a macro to make it conditionally constexpr but the benefits in xlnt are still very low

Reply:
Also have the advantage of comparing two different types without specifying a template parameter 
* detail::min(float, double) vs std::min<double>(float, double)